a refresh of the typos patch from 1.1.1 version
Index: bluegriffon-1.5.2/mozilla-2.0/db/sqlite3/src/sqlite3.c
===================================================================
--- bluegriffon-1.5.2.orig/mozilla-2.0/db/sqlite3/src/sqlite3.c	2012-11-05 20:08:59.000000000 +0100
+++ bluegriffon-1.5.2/mozilla-2.0/db/sqlite3/src/sqlite3.c	2012-11-06 10:11:56.000000000 +0100
@@ -350,7 +350,7 @@
 ** The SQLITE_THREADSAFE macro must be defined as 0, 1, or 2.
 ** 0 means mutexes are permanently disable and the library is never
 ** threadsafe.  1 means the library is serialized which is the highest
-** level of threadsafety.  2 means the libary is multithreaded - multiple
+** level of threadsafety.  2 means the library is multithreaded - multiple
 ** threads can use SQLite as long as no two threads try to use the same
 ** database connection at the same time.
 **
Index: bluegriffon-1.5.2/mozilla-2.0/gfx/gl/GLContextProviderOSMesa.cpp
===================================================================
--- bluegriffon-1.5.2.orig/mozilla-2.0/gfx/gl/GLContextProviderOSMesa.cpp	2012-11-05 20:09:00.000000000 +0100
+++ bluegriffon-1.5.2/mozilla-2.0/gfx/gl/GLContextProviderOSMesa.cpp	2012-11-06 10:16:16.000000000 +0100
@@ -131,7 +131,7 @@
     };
 
     if (!GLLibraryLoader::LoadSymbols(mOSMesaLibrary, &symbols[0])) {
-        LogMessage("Couldn't find required entry points in OSMesa libary");
+        LogMessage("Couldn't find required entry points in OSMesa library");
         return false;
     }
 
Index: bluegriffon-1.5.2/mozilla-2.0/modules/freetype2/vms_make.com
===================================================================
--- bluegriffon-1.5.2.orig/mozilla-2.0/modules/freetype2/vms_make.com	2012-11-05 20:09:02.000000000 +0100
+++ bluegriffon-1.5.2/mozilla-2.0/modules/freetype2/vms_make.com	2012-11-06 10:17:14.000000000 +0100
@@ -1096,7 +1096,7 @@
 !
 ZLIB # sys$library:libz.olb # sys$library: # zlib.h # FT_CONFIG_OPTION_SYSTEM_ZLIB
 $   write sys$output "New driver file vmslib.dat created."
-$   write sys$output "Please customize libary locations for your site"
+$   write sys$output "Please customize library locations for your site"
 $   write sys$output "and afterwards re-execute ''myproc'"
 $   goto err_exit
 $ endif
